src/Components/Endpoints/src/Binding/DefaultFormValuesSupplier.cs

```diff
@@ -1,82 +1,113 @@
+        if (!CanBind(context.ValueType, context.FormName))
         {
-            // This is not correct, but works for primtive values.
-            // Will change the interface when we add support for complex types.
-            boundValue = result;
-            return true;
+            context.SetResult(null);
         }
 
-        boundValue = valueType.IsValueType ? Activator.CreateInstance(valueType) : null;
-        return false;
+        var deserializer = _cache.GetOrAdd(context.ValueType, CreateDeserializer);
+        Debug.Assert(deserializer != null);
+        deserializer.Deserialize(context, _options, _formData.Entries, _formData.FormOptions);
     }
 
-    private Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object> CreateDeserializer(Type type) =>
-        _method.MakeGenericMethod(type)
-        .CreateDelegate<Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object>>();
+    private FormValueSupplier CreateDeserializer(Type type) =>
+        (FormValueSupplier)Activator.CreateInstance(typeof(FormValueSupplier<>)
+        .MakeGenericType(type))!;
 
-    private static object? DeserializeCore<T>(IReadOnlyDictionary<string, StringValues> form, FormDataMapperOptions options, string value)
+    internal abstract class FormValueSupplier
     {
-        // Form values are parsed according to the culture of the request, which is set to the current culture by the localization middleware.
-        // Some form input types use the invariant culture when sending the data to the server. For those cases, we'll
-        // provide a way to override the culture to use to parse that value.
-        var reader = new FormDataReader(form, CultureInfo.CurrentCulture);
-        reader.PushPrefix(value);
-        return FormDataMapper.Map<T>(reader, options);
+        public abstract void Deserialize(
+            FormValueSupplierContext context,
+            FormDataMapperOptions options,
+            IReadOnlyDictionary<string, StringValues> form,
+            FormOptions formOptions);
     }
 
-    public bool CanConvertSingleValue(Type type)
+    internal class FormValueSupplier<T> : FormValueSupplier
     {
-        return _options.IsSingleValueConverter(type);
+        public override void Deserialize(
+            FormValueSupplierContext context,
+            FormDataMapperOptions options,
+            IReadOnlyDictionary<string, StringValues> form,
+            FormOptions formOptions)
+        {
+            if (form.Count == 0)
+            {
+                return;
+            }
+
+            char[]? buffer = null;
+            try
+            {
+                var dictionary = new Dictionary<FormKey, StringValues>();
+                foreach (var (key, value) in form)
+                {
+                    dictionary.Add(new FormKey(key.AsMemory()), value);
+                }
+                buffer = ArrayPool<char>.Shared.Rent(formOptions.KeyLengthLimit);
```

Comment:
This looks potentially dangerous. Or am I misunderstanding?

Judging from https://github.com/dotnet/aspnetcore/issues/13719 it seems at least some people are in the habit of setting a really big value for this. The name of the option only implies "I accept anything up to this length" and does *not* imply "allocate this much memory regardless", so it's reasonable for people to think they can pick a big value.

Since lines 86-89 already enumerate all the keys, they could easily keep track of the largest length they've seen. Then maybe:

 * If it turns out to be < 1024 (99.9% of cases), we could stackalloc here
 * If not, but it's < formOptions.KeyLengthLimit, we could do a regular heap allocation, and hence not have to worry about cross-request contention from pooling
 * Otherwise throw

What do you think?

Reply:
> Since lines 86-89 already enumerate all the keys, they could easily keep track of the largest length they've seen. Then maybe:

This is not correct (as I discovered) when a key is not sent (for example as part of a checkbox) as we iterate based on the model, not the keys that we see, we need to be able to "hold" the longest key that the model can generate.

> If it turns out to be < 1024 (99.9% of cases), we could stackalloc here.

This precludes making the converters async in the future, which is not something we want to do. It's one of the outcomes that came from my discussions with the STJ folks, where there is a lot of work that goes on to implementing the deserializers due to the fact that they can't use async/await.  It's also not possible to use the reference we receive to query the underlying dictionary keys.

Reply:
OK, whether we stackalloc isn't really as important as being sure not to allocate massive arrays. Simply allocating based on `formOptions.KeyLengthLimit` up front looks very dangerous, unless you are saying that already happens elsewhere in the system.

Reply:
Also realistically we are likely precluded from making async converters in the future anyway, by the nature of async methods being viral and the call stack up to this point not already being async.

Having a stackalloc here would be a super minor implementation detail that would be way easier to change than getting the public API and callstack up to this point to be async.

Reply:
> Having a large formOptions.KeyLengthLimit implies that you are already willing to allocate a large value for the keys.

Only to the extent that they fit inside the max request length limit, though right? If the default request limit is 30MB, that's the longest key you could receive. Someone setting `KeyLengthLimit=int.MaxValue` would think "it can't be above 30MB" without realizing they will now allocate 2GB per request, regardless of the size of a posted form.

Rather than worry too hard about this, do you think we could get away with using something like `Math.Min(formOptions.KeyLengthLimit, 50KB)`? A pooled 50KB allocation per request is not going to bring anyone's server down, and I can't think of scenarios where people would legitimately be receiving form keys > 50KB (even that's kind of absurd).

Reply:
> Rather than worry too hard about this, do you think we could get away with using something like Math.Min(formOptions.KeyLengthLimit, 50KB)? A pooled 50KB allocation per request is not going to bring anyone's server down, and I can't think of scenarios where people would legitimately be receiving form keys > 50KB (even that's kind of absurd).

Wouldn't this be problematic?  If you explicitly allow larger keys, then the system doesn't work and you don't have a way to configure it?

Reply:
We can provide an explicit option to set a concrete limit, will that work?

Reply:
> Wouldn't this be problematic? If you explicitly allow larger keys, then the system doesn't work and you don't have a way to configure it?

That's reasonable; I was just trying to keep the scope small by excluding that case. If you think anyone would be in the situation where they have a genuine wish to receive megabytes-long form keys, that's fine with me.

> We can provide an explicit option to set a concrete limit, will that work?

I'm happy to go with your judgement on this.

Reply:
I've made it a config option on the binder that defaults to `FormReader.MaxKeyLengthLimit`. It's internal for now, but we'll expose it somewhere in an options type, once @captainsafia and I figure out the details.

Reply:
Sounds perfect.